Sprint 2 : édition et affichage d'histoires

AntoineSPR opened the pull request:
Parties narration et création fonctionnelles mais sans style
Narration : les histoires s'affichent, les scènes sont chargées suivant les choix, les dialogues défilent.
Création : CRUD des histoires et des scènes, avec personnages, dialogues et choix

Kwett committed:
Add Story Creation Form

Kwett committed:
Add scene creation form

AntoineSPR committed:
Renamed files, classes and routes

AntoineSPR committed:
Removed unnecessary method and fixed whitespace

AntoineSPR committed:
Added story and scene name on scene creation screen

AntoineSPR committed:
Added list of scenes on the story creation screen

AntoineSPR committed:
Fixed whitespaces

Commit:
feat: Add StoryController and StoryManager to handle story selection

- Added StoryController to handle the logic for displaying the story selection page.
- Added StoryManager to retrieve the list of stories from the database.
- Created a new template file, index.html.twig, for the story selection page.
- Updated routes.php to include the '/story' route and map it to the StoryController's index method.

✨ feat(Ajout de la gestion de l'affiche des histoires controlleurs et modèles associés, mise en place des vues twig -> sans css pour l'instant ni images de personnages / backgrounds):

Co-authored-by: Stephie9233 <Stephie9233@users.noreply.github.com>

Commit:
refactor: Update variable names and fix method calls in SceneController

- Update variable names in SceneController to follow camel case convention.
- Fix method calls in SceneController to use the updated variable names.
- Remove commented out code and unused imports.

Co-authored-by: Stephie9233 <Stephie9233@users.noreply.github.com>

AntoineSPR committed:
Added link back to story in scene screen and fixed method

AntoineSPR committed:
Fixed twig parameters

AntoineSPR committed:
Added link to each scene in story creation screen & fixed methods

AntoineSPR committed:
Added index for story creation screen

AntoineSPR committed:
Added link back to stories index from story creation screen

AntoineSPR committed:
Fixed whitespace

Kwett committed:
Add deletions link

Kwett committed:
Merge branch 'S1US2_story_creation' of github.com:WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite into S1US2_story_creation

AntoineSPR committed:
Modified methods and routes before merging with US1

Commit:
trop de trucs de faits manque : le css

Commit:
refactor: Update variable names and fix method calls in SceneController

- Update variable names in SceneController to follow camel case convention.
- Fix method calls in SceneController to use the updated variable names.
- Remove commented out code and unused imports.

Commit:
tailwind to normal css

Commit:
fix css

Commit:
refactor: Update database seed script to include new story and scenes

- Add a new story "L'énigme de la forêt enchantée" to the seed script.
- Insert scenes for the new story, including background images.
- Update character and dialogue line data to match the new scenes.

Commit:
🌈 style(Mise à jour du css):

Commit:
🌈 style(css fix buttons):

Commit:
fix css & page title in twig for scenes
todo-> fix page title for homepages

AntoineSPR committed:
Added dialogues and choices to scene creation screen

AntoineSPR committed:
Fixed links

AntoineSPR committed:
Fixed links

Kwett committed:
add character deletion

Kwett committed:
fix database

Kwett committed:
fix database

Commit:
🐞 fix: Simplficiation du code

- Refonte de la vue pour une scène et simplificatoin du code JS
- Suppression des fichiers CSS
- Retrait de commentaires inutiles dans les controlleurs
- Ajout de commentaires

Commit:
🐞 fix: Simplficiation du code

- Refonte de la vue pour une scène et simplificatoin du code JS
- Suppression des fichiers CSS
- Retrait de commentaires inutiles dans les controlleurs
- Ajout de commentaires

Commit:
✨ feat(mise en place de la redirection vers la page de crédits):

AntoineSPR committed:
Added dialogue creation screen

AntoineSPR committed:
Fixed whitespaces

AntoineSPR committed:
Fixed dialogue list

AntoineSPR committed:
Added character edit

AntoineSPR committed:
Added functionalities to character edit

AntoineSPR committed:
Refactored, fixed special characters and prevented injection

AntoineSPR committed:
Merge branch 'S1US1_affichage_general' into S2_dev_temp

Commit:
🐞 fix(fix method calls in SceneController):

Commit:
🐞 fix(grumPHP = 💩):

Kwett committed:
fix html decode

Co-authored-by: AntoineSPR <AntoineSPR@users.noreply.github.com>

AntoineSPR committed:
Merge branch 'S1US1_affichage_general' into S2_dev_temp

AntoineSPR committed:
Added choices CRUD

AntoineSPR committed:
Cleaned up twig view for scene creation screen

AntoineSPR committed:
Moved dialogue form on scene creation screen

AntoineSPR committed:
Fixed forms on scene creation screen

AntoineSPR committed:
Fixed layout for scene creation screen

AntoineSPR committed:
Fixed database encoding and seed

AntoineSPR committed:
Added script to scene creation screen

AntoineSPR committed:
Fixed link at the end of a story

AntoineSPR committed:
Updated composer files

AntoineSPR committed:
Fixed composer.lock

AntoineSPR committed:
image upload

Commit:
chore: Update character sprite upload functionality

Co-authored-by: Kwett <Kwett@users.noreply.github.com>

Commit:
chore: Update character sprite upload functionality

Commit:
feat: Update background image styling in show.html.twig

Commit:
🎵 composer est composed

Commit:
Composer nouvel essai

Commit:
Update GitHub Actions workflow to use PHP 8.2

Commit:
🐳 chore:

.................. ../¯/) .
. ................. /¯../
. ................/... ./
. ........./;¯/;...;/; ¯`•;
. ......./;/.../..../. ....../;¯\
. .....(....;....;.... . ~/;...;)
. .......\. ............. .;....(
. ........\......... ...... _.•;
. ..........\ .................(
. .............\... ..........\

mickaellambert commented on .github/workflows/main.yml:
Pas besoin de modifier ce fichier, c'est pour le déploiement du site

Amine03824 replied:
C'était obligatoire de mettre la nouvelle version de php sinon depuis hier soir plus rien ne passait les checks
![image](https://github.com/WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite/assets/137644798/09f49ac7-2363-43b7-9f3c-878d2dd88605)
![image](https://github.com/WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite/assets/137644798/9a64f2cd-1b09-4fe0-a7a9-c6e1e5b19d41)

le principal changement est au niveau de la version php pour github actions 
![image](https://github.com/WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite/assets/137644798/43b05ac2-b226-4969-93ba-098b11561291)

mickaellambert commented on docker-compose.override.yml:
Pas besoin de modifier ce fichier, c'est pour le déploiement du site

mickaellambert commented on docker-compose.prod.yml:
Pas besoin de modifier ce fichier, c'est pour le déploiement du site

mickaellambert commented on docker-compose.yml:
Pas besoin de modifier ce fichier, c'est pour le déploiement du site

mickaellambert commented on public/assets/js/scene.js:
Attention à ne pas polluer votre code de commentaires inutiles, il est censé être compréhensible sans

mickaellambert commented on src/Controller/AbstractController.php:
L'idée est bonne, mais malheureusement pas forcément la plus optimale. A chaque instanciation de controller, vous allez instancier plusieurs managers qui ne seront pas forcément utilisé dans les dits controller.

mickaellambert commented on src/Controller/AbstractController.php:
```diff
@@ -5,13 +5,25 @@
+    protected CharacterManager $characterManager;
+    protected StoryManager $storyManager;
+    protected ChoiceManager $choiceManager;
+    public const EXTENSIONS_ALLOWED = ['jpg', 'jpeg', 'png', 'webp', 'svg'];
```
Ici, la constante n'est pas bien placé, est ce le type de fichier accepté pour le formulaire d'inscription et l'avatar, est ce le type de fichier accepté pour la création de dialogue, de sprite ? Je pourrai trouver d'autres idées mais globalement il vaut mieux placer cette constante dans le controlleur qu'elle concerne directement (et la nommer en conséquence)

mickaellambert commented on src/Controller/CharacterController.php:
```diff
@@ -0,0 +1,70 @@
+
+            if (!move_uploaded_file($_FILES['sprite']['tmp_name'], $targetFile)) {
+                $errors[] = 'Erreur lors du déplacement du fichier de sprite';
+                // Log des informations d'erreur
```
Commentaire inutile

mickaellambert commented on src/Controller/CharacterController.php:
```diff
@@ -0,0 +1,70 @@
+                $this->characterManager->insert($character);
+            } else {
+                // Log des erreurs dans un fichier de journal
+                error_log('Erreurs lors de l\'ajout dtu sprite: ' . implode(', ', $errors));
```
Très bien de faire ça bravo

mickaellambert commented on src/Controller/CharacterController.php:
```diff
@@ -0,0 +1,70 @@
+                $errors[] = 'Votre sprite n\'as pas le bon format (' . implode(', ', parent::EXTENSIONS_ALLOWED) . ')';
+            }
+
+            if (!move_uploaded_file($_FILES['sprite']['tmp_name'], $targetFile)) {
```
Ne voulez vous pas réaliser l'upload du fichier que dans le cas ou il n'y a pas d'erreurs sur votre formulaire ?

mickaellambert commented on src/Controller/CharacterController.php:
```diff
@@ -0,0 +1,70 @@
+    {
+        if ($_SERVER['REQUEST_METHOD'] === 'POST') {
+            $character = array_map('htmlentities', array_map('trim', $_POST));
+            $character['name'] = $character['e_name'];
```
Pourquoi ne pas définir les attributs name de vos input directement à "name" et "sprite" ?

mickaellambert commented on src/Controller/ChoiceController.php:
```diff
@@ -0,0 +1,40 @@
+        if ($_SERVER['REQUEST_METHOD'] === 'POST') {
+            $choice = array_map('htmlentities', array_map('trim', $_POST));
+            $choice['scene_id'] = $sceneId;
+            $choice['next_scene_id'] = $nextSceneId;
```
Pourquoi le next_scene_id est envoyé en GET et pas en POST ?

mickaellambert commented on src/Controller/SceneController.php:
```diff
@@ -0,0 +1,138 @@
+            $scene = array_map('htmlentities', array_map('trim', $_POST));
+            $scene['story_id'] = $storyId;
+
+            $targetDir = 'assets/images/backgrounds/';
```
Cette valeur pourrait être une constante dans votre controller

mickaellambert commented on src/Controller/SceneController.php:
```diff
@@ -0,0 +1,138 @@
+
+            if (!move_uploaded_file($_FILES['background']['tmp_name'], $targetFile)) {
+                $errors[] = 'Erreur lors du déplacement du fichier de background';
+                // Log des informations d'erreur
```
Même commentaire que précédemment, faites attentions aux commantaires de code inutile

mickaellambert commented on src/Controller/SceneController.php:
```diff
@@ -0,0 +1,138 @@
+        }
+
+        // Utiliser DialogueManager pour obtenir les dialogues de la scène
+        $dialogues = $this->dialogueManager->getDialoguesBySceneId($sceneId);
```
Essayez de rester cohérents dans vos manager : les fonctions de selection de base commencent par le préfixe "select" et les votre par "get", ca peut rendre la chose plus confusante pour des développements futurs

Pas besoin de préciser la ressource concernée dans le nom, nous sommes déjà dans un dialogueManager.

Je verrai plus "selectAllByScene"

mickaellambert commented on src/Controller/SceneController.php:
```diff
@@ -0,0 +1,138 @@
+        $dialogues = $this->dialogueManager->getDialoguesBySceneId($sceneId);
+
+        // Utiliser ChoiceManager pour obtenir les choix de la scène
+        $choices = $this->choiceManager->getChoicesBySceneId($sceneId);
```
selectAllByScene

mickaellambert commented on src/Controller/SceneController.php:
```diff
@@ -0,0 +1,138 @@
+     * faire une condition qui changerait la requête en ajoutant ORDER BY id ASC LIMIT 1
+     */
+
+    public function showFirstScene(int $storyId): string
```
Pourquoi existe-t-il une route qui permet d'afficher la première scène ? 
Il faudrait, lorsque vous cliquez sur le bouton "commencer une histoire" que vous fournissiez le lien vers la route show de base d'une scène avec l'id de la première scène. 
Ne pas ajouter un comportement différent dans un cas précis

mickaellambert commented on src/Controller/SceneController.php:
```diff
@@ -0,0 +1,138 @@
+
+    public function showFirstScene(int $storyId): string
+    {
+        $firstSceneId = $this->sceneManager->findFirstSceneIdOfStory($storyId);
```
La fonction devrait plutôt s'appeler "selectFirstByStory" et devrait retourner toutes les données d'une scène (sait on jamais, vous en aurez peut être besoin un jour)

mickaellambert commented on src/Controller/SceneController.php:
```diff
@@ -0,0 +1,138 @@
+        }
+
+        // Afficher la première scène de l'histoire
+        return $this->show($storyId, $firstSceneId);
```
Du coup ici : 

`return $this->show($storyId, $firstScene['id']);`

mickaellambert commented on src/Controller/StoryController.php:
```diff
@@ -0,0 +1,57 @@
+
+class StoryController extends AbstractController
+{
+    public function indexCreation(): ?string
```
J'aurai mis les routes concernant les création d'histoires dans un autre controller "StoryCreation" ou même "Backoffice"

mickaellambert commented on src/Model/AbstractManager.php:
Il ne faut pas toucher à ce fichier, le framework marche très bien comme il est

mickaellambert commented on src/Model/AbstractManager.php:
```diff
@@ -56,4 +57,18 @@ public function delete(int $id): void
         $statement->execute();
     }
+
+    public function decodeHtmlEntitiesInArray(array|null $array): ?array
```
Je comprends pas trop le probleme que vous essayez de résoudre ici

mickaellambert commented on src/Model/DialogueManager.php:
```diff
@@ -0,0 +1,65 @@
+        return $statement->execute();
+    }
+
+    public function getDialogues(string $id): ?array
```
Vous pouvez surcharger la fonction selectAll() présente dans l'Abstract

mickaellambert commented on src/Model/DialogueManager.php:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Récupère les dialogues & les personnages associés par l'ID de la scène.
+     */
+    public function getDialoguesBySceneId(int $sceneId): ?array
```
Voir commentaire plus haut

mickaellambert commented on src/Model/SceneManager.php:
```diff
@@ -0,0 +1,74 @@
+        return (int) $this->pdo->lastInsertId();
+    }
+
+    public function getStory(string $id): ?array
```
Pourquoi rajouter une fonction qui récupère une histoire dans un SceneManager ? 
De plus, il existe déjà la fonction selectOneById sur tout vos managers

mickaellambert commented on src/Model/SceneManager.php:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * Récupère scène par son ID et celui de l'histoire.
+     */
+    public function findScene(int $storyId, int $sceneId): array|false
```
Pourquoi ajouter le storyId dans vos critères de recherches ? Cela marcherait très bien avec juste l'ID de la scene et donc pas besoin de créer une nouvelle fonction

mickaellambert commented on src/Model/SceneManager.php:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * Récupère l'ID de la première scène associée à une histoire.
+     */
+    public function findFirstSceneIdOfStory(int $storyId): int|false
```
Voir commentaire plus haut

mickaellambert commented on src/Model/SceneManager.php:
```diff
@@ -0,0 +1,74 @@
+        return intval($row['id']);
+    }
+
+    public function selectAllByStoryId(string $storyId): array
```
selectAllByStory()

mickaellambert commented on src/Model/StoryManager.php:
```diff
@@ -0,0 +1,48 @@
+
+class StoryManager extends AbstractManager
+{
+    public const TABLE = "`story`";
```
Ca ne marche pas comme ça ? `public const TABLE = "story";`

mickaellambert commented on src/Model/StoryManager.php:
```diff
@@ -0,0 +1,48 @@
+        return $statement->execute();
+    }
+
+    public function getScenes(string $storyId): ?array
```
selectScenes

mickaellambert commented on src/Model/StoryManager.php:
```diff
@@ -0,0 +1,48 @@
+        return $this->decodeHtmlEntitiesInArray($scenes);
+    }
+
+    // Hérite de la méthode SelectAll pour récupérer toutes les histoires après défintion de la table
```
commentaire inutile

mickaellambert commented on src/routes.php:
```diff
@@ -7,5 +7,26 @@
+        'SceneController', 'showFirstScene', ['storyId'] // Affiche la première scène de l'histoire
+    ],
+    'scene/show' => ['SceneController', 'show', ['storyId', 'sceneId']], // Affiche une scène à l'aide de son histoire
+    'storycreation' => ['StoryController', 'indexCreation'],
```
Ca me parait plus cohérent de faire des routes ressemblant à ça : /story/admin/

Pas obligé, qu'en pensez vous ?

mickaellambert commented on src/routes.php:
```diff
@@ -7,5 +7,26 @@
+    'storycreation/scene/add_dial' => ['DialogueController', 'add', ['story_id', 'scene_id']],
+    'storycreation/scene/del_dial' => ['DialogueController', 'delete', ['story_id', 'scene_id', 'id']],
+    'storycreation/scene/update_dial' => ['DialogueController', 'update', ['story_id', 'scene_id', 'id']],
+    'storycreation/addchar' => ['CharacterController', 'add', ['story_id', 'scene_id']],
```
Si vous faites des routes en séparant les mots par des "_" au dessus, pensez bien à le faire partout.

mickaellambert commented on src/routes.php:
```diff
@@ -7,5 +7,26 @@
+    'storycreation/scene/add' => ['SceneController', 'add', ['story_id']],
+    'storycreation/scene/show' => ['SceneController', 'showCreation', ['story_id', 'id']],
+    'storycreation/scene/delete' => ['SceneController', 'delete', ['story_id', 'id']],
+    'storycreation/scene/add_dial' => ['DialogueController', 'add', ['story_id', 'scene_id']],
```
Que pensez vous de `story/admin/scene/dialogue/add` comme format global de routes ?

mickaellambert commented on src/routing.php:
Attention aux fichiers que vous envoyez avec des modifs toute simple comme celle ci

mickaellambert reviewed (changes requested):
Ne vous faites pas avoir par le nombre de commentaires, c'est globalement du très bon travail, vous avez bien organisé votre code, tenter des choses et vous avez bien compris comment l'architecture MVC marche. 

Bravo pour tout ça !

Kwett committed:
fixed managers instanciation

Kwett committed:
fix upload const position

Kwett committed:
fix targetDir as a constant

Co-authored-by: Amine <Amine03824@users.noreply.github.com>
Co-authored-by: AntoineSPR <AntoineSPR@users.noreply.github.com>

Kwett committed:
Fix routes and method's name

Kwett committed:
fix findFirstScene method's name (selectFirstByStory)

Commit:
🐞 fix(ce qui était bugged est fixed):

- Récupération de toute la table character dans selectAllByScene du DialogueManager afin d'afficher correctement les sprites
- Résolution d'un bug de de showFirstScene du sceneController résultant dans l'affichage de la scène 2 de l'histoire 1 au lieu de la scène 1 de l'histoire 2 🤔

Stephie9233 committed:
CSS - First part

Kwett committed:
Replace Get data by Post
Rename array_keys for forms

Co-authored-by: AntoineSPR <AntoineSPR@users.noreply.github.com>

Kwett committed:
Merge branch 's2_fix_pr1' of github.com:WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite into s2_fix_pr1

AntoineSPR committed:
Merge branch 's2_fix_pr1' of github.com:WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite into S3_dev_temp

AntoineSPR committed:
Updated database and corrected a Twig property

AntoineSPR committed:
Renamed all routes and fixed bugs in scene creation screen

Commit:
🐞 fix:

refactor: Improve dialogue navigation logic in scene.js  The code changes in scene.js refactor the dialogue navigation logic to improve readability and maintainability. The functions `showNextDialogue` and `checkIfLastDialogue` have been modified to handle the display of dialogues and adapt the UI based on the current dialogue index. Additionally, the function `checkIfLastDialogue` now correctly updates the "continue" button text to "Fin" when it is the last dialogue. This change enhances the user experience and ensures a smooth flow through the dialogues.

AntoineSPR committed:
Added confirmation message upon deletion (story and scene)

AntoineSPR committed:
Merge branch 'S3_client_fix' into S3_dev_temp

Kwett committed:
adds validation on forms
adds autofill when empty forms

Kwett committed:
Merge branch 'S3_dev_temp' of github.com:WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite into S3_form_validation

AntoineSPR committed:
Added the option to create a choice without path

AntoineSPR committed:
Added cancel button upon adding elements in scene

Kwett committed:
add default value on forms

Kwett committed:
Merge branch 'S3_form_validation' of github.com:WildCodeSchool-2024-02/php-poec-paris-p2-herowebsite into S3_form_validation

AntoineSPR committed:
Fixed update and added one for scene name

AntoineSPR committed:
Merge branch 'S3_client_fix' into S3_dev_temp

AntoineSPR committed:
Deleted unnecessary Controller

AntoineSPR committed:
Add raw filter, fix controllers names, override SelectAll
